Apache Ignite Dataset

dmitrievanthony opened the pull request:
This is a proposal to add `IgniteDataset` that allows to work with [Apache Ignite](https://ignite.apache.org/). 

Apache Ignite is a memory-centric distributed database, caching, and processing platform for transactional, analytical, and streaming workloads, delivering in-memory speeds at petabyte scale. This proposal is a part of a more global initiative to implement so called "TensorFlow on Apache Ignite" ([IGNITE-8335](https://issues.apache.org/jira/browse/IGNITE-8335), [Design Document](https://docs.google.com/document/d/1jROIahK1rc7bSgOvhJhfpMqIGvht_IE8zn5NAt6x8ks)).

The integration is based on Apache Ignite [Binary Client Protocol](https://apacheignite.readme.io/v2.6/docs/binary-client-protocol) and TensorFlow [tf.data.Dataset](https://www.tensorflow.org/guide/datasets). More information about supported features you can find in [README.md](https://github.com/dmitrievanthony/tensorflow/blob/apache-ignite-dataset/tensorflow/contrib/ignite/README.md) of this module.

Tests have also been added. They use docker to hide configuration complexity, so that the implemented functionality can be tested quite simply by manual run.

dmitrievanthony committed:
Merge pull request #1 from tensorflow/master

42

dmitrievanthony committed:
Merge pull request #3 from tensorflow/master

42

dmitrievanthony committed:
Add IgniteDataset that allows to work with Apache Ignite.

dmitrievanthony committed:
Remove duplicated header from README.md.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.h:
```diff
@@ -0,0 +1,87 @@
+#include "ignite_binary_object_parser.h"
+#include "ignite_dataset.h"
+
+#ifndef IGNITE_CLIENT_H
```
For header guards like this one, please use the full path (and this one seems not to match the filename):

```c++
#ifndef TENSORFLOW_CONTRIB_IGNITE_KERNELS_IGNITE_DATASET_ITERATOR_H_`
```

dmitrievanthony replied:
Fixed.

mrry commented on tensorflow/contrib/ignite/kernels/ignite_dataset_ops.cc:
```diff
@@ -0,0 +1,145 @@
+#include "tensorflow/core/framework/dataset.h"
+
+namespace tensorflow {
+
```
Please wrap this class and the registration in an anonymous namespace.

dmitrievanthony replied:
Fixed.

mrry reviewed (changes requested):
Thanks for resubmitting this PR. Since the PR is extra-large, it may take several weeks to review in its entirety. I've taken a quick pass and made some comments on style that apply to all of the code, so please apply those first.

dmitrievanthony committed:
Update after review: change 'ignite' namespace to 'tensorflow', rename variables to satisty code style, use pointers instead of references.

dmitrievanthony commented:
Thank you for reviewing, @mrry. Really hope it won't take more than two weeks to review and make this request merge-ready because in Apache Ignite we have big plans about it. Feel free to contact me any time if you have questions about this PR.

I've updated the code in accordance with all comments you have so far, so I would be grateful if you could continue your review.

dmitrievanthony committed:
Update README.md.

dmitrievanthony commented:
Hi, @mrry, @martinwicke. Any updates?

By the way, could you please run CI tests so that I am able to fix now if something wrong?

dmitrievanthony commented:
Guys, @mrry, @martinwicke, I really understand that you are under pressure of all these pull requests, but it's been more than 10 days since your last comment here, so let me ping you and ask to take a look.

martinwicke commented:
I've retriggered the tests. Sorry for the wait.

dmitrievanthony committed:
Fix pylint checks, fix VS compilation issue.

dmitrievanthony commented:
Hi, @martinwicke. Thank you for restarting tests. Sorry for bothering, but I still can't fully reproduce CI build.

I see that during the last run only one required test failed (one `pylint` check to be precise). I have just fixed it, so could you please restart tests again?

And one more question. Should only `required` tests be green or all? I see that a lot of "none-required" tests fail in absolutely another modules with timeout and other types of exceptions, so I don't feel like I can fix them.

martinwicke commented:
@mrry can you take a look at the changes?

dmitrievanthony commented:
Hi, @mrry, @martinwicke. I feel yourself a bit lost here. It's been 17 days since last @mrry comment and the PR is on the 3rd page already. Guys, could you please pay attention to this request?

perfinion commented:
@dmitrievanthony sorry for all the delay. I re-triggered the tests, looks like there is some issue with the windows build and clang-format (The XLA tests fail a lot, I'm not sure they are related), can you take a look?

Also can you rebase and fix up the commits? The Merge Commits in the PR are weird i dont follow them. If you do an interactive rebase on origin/master you can drop them and fix up the commits.

I'm not really familiar with these areas of TF so can't review all of it but I'll do what I can.

dmitrievanthony committed:
Fix code style.

dmitrievanthony commented on tensorflow/contrib/ignite/__init__.py:
```diff
@@ -0,0 +1,42 @@
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.contrib.ignite.python.ops.ignite_dataset_ops \
```
Fixed.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,304 @@
+
+#include "ignite_binary_object_parser.h"
+
+namespace ignite {
```
Fixed.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,304 @@
+                                tensorflow::DT_UINT8, {});
+      tensor.scalar<tensorflow::uint8>()() = *((uint8_t*)ptr);
+      ptr += 1;
+      out_tensors.emplace_back(std::move(tensor));
```
Fixed.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc:
```diff
@@ -0,0 +1,304 @@
+namespace ignite {
+
+tensorflow::Status BinaryObjectParser::Parse(
+    uint8_t*& ptr, std::vector<tensorflow::Tensor>& out_tensors,
```
Fixed.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_client.cc:
```diff
@@ -0,0 +1,55 @@
+
+namespace ignite {
+
+tensorflow::Status Client::ReadByte(uint8_t& data) {
```
Fixed.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset.h:
```diff
@@ -0,0 +1,65 @@
+      tensorflow::Node** output) const override;
+
+ private:
+  const std::string cache_name;
```
Fixed.

dmitrievanthony commented on tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc:
```diff
@@ -0,0 +1,447 @@
+
+namespace ignite {
+
+#define CHECK_STATUS(status) \
```
Fixed.

dmitrievanthony commented:
Hi @perfinion. I've fixed code style and removed "merge requests", but as result I've recreated the branch, so could you please continue with #22210?

Regarding builds on Windows I see the same issues I had and asked here: https://groups.google.com/a/tensorflow.org/forum/#!topic/build/ePYss0Kxcu4. It looks like we need to add `copt=-DWIN32_LEAN_AND_MEAN` on CI server for Windows builds. What do you think?

perfinion commented:
oh, you could have just force-pushed to this branch name. but sure, I've run tests and re-assigned that PR instead.
